deepgram/clients/abstract_restful_client.py

```diff
@@ -0,0 +1,81 @@
+import httpx
+import json
+from typing import Dict, Any, Optional
+from ..errors import DeepgramApiError, DeepgramUnknownApiError, DeepgramUnknownError
+from ..helpers import is_json
+
+class AbstractRestfulClient:
+    """
+    An abstract base class for a RESTful HTTP client.
+
+    This class provides common HTTP methods (GET, POST, PUT, PATCH, DELETE) for making asynchronous HTTP requests.
+    It handles error responses and provides basic JSON parsing.
+
+    Args:
+        url (Dict[str, str]): The base URL for the RESTful API, including any path segments.
+        headers (Optional[Dict[str, Any]]): Optional HTTP headers to include in requests.
+
+    Attributes:
+        url (Dict[str, str]): The base URL for the RESTful API.
+        client (httpx.AsyncClient): An asynchronous HTTP client for making requests.
+        headers (Optional[Dict[str, Any]]): Optional HTTP headers to include in requests.
+
+    Exceptions:
+        DeepgramApiError: Raised for known API errors.
+        DeepgramUnknownApiError: Raised for unknown API errors.
+        DeepgramUnknownError: Raised for unexpected errors not specific to the API.
+    """
+    def __init__(self, url: Dict[str, str], headers: Optional[Dict[str, Any]]):
+        self.url = url
+        self.client = httpx.AsyncClient()
+        self.headers = headers
+
+    async def get(self, url: str, options=None):
+        headers = self.headers
+        return await self._handle_request('GET', url, params=options, headers=headers)
+
+    async def post(self, url: str, options=None, **kwargs):
+        headers = self.headers
+        return await self._handle_request('POST', url, params=options, headers=headers, **kwargs)
+    
+    async def put(self, url: str, options=None, **kwargs):
+        headers = self.headers
+        return await self._handle_request('PUT', url, params=options, headers=headers, **kwargs)
+
+    async def patch(self, url: str, options=None, **kwargs):
+        headers = self.headers
+        return await self._handle_request('PATCH', url, params=options, headers=headers, **kwargs)
+
+    async def delete(self, url: str):
+        headers = self.headers
+        return await self._handle_request('DELETE', url, headers=headers)
+
+    async def _handle_request(self, method, url, **kwargs):
+        retries = 0
+        max_retries = 4
+        while retries < max_retries:
```

Comment:
This was something that was in the v2 SDK. Luke also suggested I remove it. I added this because at one point I hit a timeout error and I thought maybe it was worth keeping.

But I think you're right to question this. In the SDKs meeting we discussed adding a feature that gives the user the power to choose if there is a retry. I think I'll remove this and then we can create the issue or RFC to discuss if we want to add a feature for the user to add the retry

Reply:
Sounds fair. Usually retries are typical for establishing a persistent connection, and are "usually" atypical for single RPC, REST, etc calls.